add Expo config plugin

kbrandwijk opened the pull request:
## :loudspeaker: Type of change
<!--- Put an `x` in the boxes that apply -->
- [ ] Bugfix
- [X] New feature
- [ ] Enhancement
- [ ] Refactoring


## :scroll: Description
<!--- Describe your changes in detail -->
This PR adds the Expo config plugin that used to be part of the `sentry-expo` package.
The plugin implementation is identical, except for moving the configuration properties over from the (deprecated) `postPublish` configuration in app.json to config plugin properties that is the default for config plugins.

## :bulb: Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Closes https://github.com/getsentry/sentry-react-native/issues/3263
For more details, see the linked issue and the parent issue.

## :green_heart: How did you test it?
Tested by building and testing the config plugin locally, and adding it to a local Expo test project to see if the config plugin was applied successfully.

## :pencil: Checklist
<!--- Put an `x` in the boxes that apply -->
- [X] I reviewed submitted code
- [X] I added tests to verify changes
- [X] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled
- [X] All tests passing
- [X] No breaking changes

## :crystal_ball: Next steps
:shipit:

krystofwoldrich commented:
@kbrandwijk Thank you, this is a good start. There are a few minor things we'll need to adjust.

I've changed the target to `expo` branch so we can get all changes there before merging to `main` and releasing.

I would also like to add a simple Expo app to test a build with the plugin in CI. But we can do that in another PR.

krystofwoldrich commented on package.json:
```diff
@@ -81,6 +88,8 @@
     "eslint": "^7.6.0",
     "eslint-plugin-react": "^7.20.6",
     "eslint-plugin-react-native": "^3.8.1",
+    "expo": "^47.0.0",
+    "expo-module-scripts": "^3.1.0",
```
Could we only add `@expo/config-plugins`? 

I believe that is the only dependency used in the plugin.

kbrandwijk replied:
We recommend against adding `@expo/config-plugins` directly, because it's very easy to create version conflicts with it.

krystofwoldrich replied:
Okay, I see let's keep `expo`.

krystofwoldrich commented on package.json:
```diff
@@ -18,12 +18,16 @@
     "downlevel": "downlevel-dts dist ts3.8/dist --to=3.8",
     "clean": "rimraf dist coverage",
+    "clean:plugin": "expo-module clean plugin",
+    "prepare": "expo-module prepare",
+    "prepublishOnly": "expo-module prepublishOnly",
```
These commands should not be necessary. At the moment the `yarn install` fails on the `prepare`. 

I think we can leave them out.

kbrandwijk replied:
Removed

krystofwoldrich commented on package.json:
```diff
@@ -18,12 +18,16 @@
     "build:sdk": "tsc -p tsconfig.build.json",
     "build:sdk:watch": "tsc -p tsconfig.build.json -w --preserveWatchOutput",
     "build:tools": "tsc -p tsconfig.build.tools.json",
+    "build:plugin": "expo-module build plugin",
```
```suggestion
    "build:plugin": "EXPO_NONINTERACTIVE=true expo-module build plugin",
```

Let's keep it without watchdog by default like the other commands.

krystofwoldrich commented on package.json:
```diff
@@ -33,10 +37,12 @@
     "lint": "yarn lint:eslint && yarn lint:prettier",
     "lint:eslint": "eslint --config .eslintrc.js .",
     "lint:prettier": "prettier --check \"{src,test,scripts}/**/**.ts\"",
+    "lint:plugin": "expo-module lint plugin",
```
I think this might not be needed. We can add the plugin to `{projectRoot}tsconfig.json`

```
    "plugin/**/*.ts",
```

and then if needed any exception or lint changes can go to `.eslintrc.js`.

krystofwoldrich commented:
We can move the test to `/test/expo-plugin/*` and change the naming from `-test.ts` to `.test.ts`. 

We keep all tests in `/test`.

krystofwoldrich commented:
To keep the PR small and easy to understand I created some follow-up tasks.

- https://github.com/getsentry/sentry-react-native/issues/3393

kbrandwijk commented:
> @kbrandwijk So with this PR the configuration of the plugin would look like this:
> 
> `app.config.js`
> 
> ```js
> const { withSentry } = require('@sentry/react-native/plugin/build/withSentry');
> 
> const config = {
> 	name: 'my app',
> };
> 
> module.exports = withSentry(config, {
> 	url: 'https://example.sentry.url/',
> 	authToken: 'example-token',
> 	project: 'project-slug',
> 	organization: 'org-slug',
> });
> ```
> 
> OR in JSON
> 
> ```json
> {
> 	"name": "my app",
> 	"plugins": [
> 		[
> 			"@sentry/react-native/plugin/build/withSentry",
> 			{
> 				"url": "https://example.sentry.url/",
> 				"authToken": "example-token",
> 				"project": "project-slug",
> 				"organization": "org-slug"
> 			}
> 		]
> 	]
> }
> ```

Correct

kbrandwijk commented:
Thank you for the PR feedback. After applying the lint suggestion, I addressed most of the issues, but was left with a few nasty ones that probably require a bit more work to fix (around the xcode types for example). I disabled those for now, I suggest we circle back to these in a follow up ticket.

krystofwoldrich commented:
@kbrandwijk Sounds good. Thank you.

kbrandwijk commented:
> > @kbrandwijk So with this PR the configuration of the plugin would look like this:
> > `app.config.js`
> > ```js
> > const { withSentry } = require('@sentry/react-native/plugin/build/withSentry');
> > 
> > const config = {
> > 	name: 'my app',
> > };
> > 
> > module.exports = withSentry(config, {
> > 	url: 'https://example.sentry.url/',
> > 	authToken: 'example-token',
> > 	project: 'project-slug',
> > 	organization: 'org-slug',
> > });
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > OR in JSON
> > ```json
> > {
> > 	"name": "my app",
> > 	"plugins": [
> > 		[
> > 			"@sentry/react-native/plugin/build/withSentry",
> > 			{
> > 				"url": "https://example.sentry.url/",
> > 				"authToken": "example-token",
> > 				"project": "project-slug",
> > 				"organization": "org-slug"
> > 			}
> > 		]
> > 	]
> > }
> > ```
> 
> Correct

Actually, for end user purposes, it would just be:
```json
{
	"name": "my app",
	"plugins": [
		[
			"@sentry/react-native",
			{
				"url": "https://example.sentry.url/",
				"authToken": "example-token",
				"project": "project-slug",
				"organization": "org-slug"
			}
		]
	]
}
```

brentvatne commented on plugin/src/withSentry.ts:
```diff
@@ -0,0 +1,70 @@
+      cfg = withSentryAndroid(cfg, sentryProperties);
+    } catch (e) {
+      WarningAggregator.addWarningAndroid(
+        'sentry-expo',
```
it looks like we left some sentry-expo references in here, should update to this lib name

brentvatne commented on plugin/src/withSentryAndroid.ts:
```diff
@@ -0,0 +1,54 @@
+  if (!buildGradle.match(pattern)) {
+    WarningAggregator.addWarningAndroid(
+      'sentry-expo',
+      'Could not find react.gradle script in android/app/build.gradle. Please open a bug report at https://github.com/expo/sentry-expo.',
```
we should replace this url with https://github.com/getsentry/sentry-react-native

objectiveSee commented:
Any update on when this release will be stable? We are setting up Sentry for our app that uses Expo 49 and I am having issues with source maps when using [Expo's sentry library](https://github.com/expo/sentry-expo). If this alpha is almost ready then we might just wait to use this instead.

krystofwoldrich commented:
@objectiveSee We are planning two releases before the stable, which should come in the upcoming weeks.
